beacon_node/beacon_chain/src/attestation_rewards.rs

```diff
@@ -0,0 +1,183 @@
+
+            for effective_balance_eth in 0..=32 {
+                let base_reward =
+                    effective_balance_eth.safe_mul(base_reward_per_increment.as_u64())?;
+
+                let penalty =
+                    !0 * base_reward.safe_mul(weight)?.safe_div(WEIGHT_DENOMINATOR)? as i64;
+
+                let reward_numerator = base_reward
+                    .safe_mul(weight)?
+                    .safe_mul(unslashed_participating_increments)?;
+
+                let ideal_reward = reward_numerator
+                    .safe_div(active_increments)?
+                    .safe_div(WEIGHT_DENOMINATOR)?;
+                if !state.is_in_inactivity_leak(previous_epoch, spec) {
+                    ideal_rewards_hashmap
+                        .insert((flag_index, effective_balance_eth, penalty), ideal_reward);
+                } else {
+                    ideal_rewards_hashmap.insert((flag_index, effective_balance_eth, penalty), 0);
+                }
+            }
+        }
+
+        // Calculate total rewards
+        let mut total_rewards: Vec<TotalAttestationRewards> = Vec::new();
+
+        if validators.is_empty() {
+            validators = participation_cache.eligible_validator_indices().to_vec();
+        }
+
+        for validator_index in &validators {
+            let eligible = state.is_eligible_validator(previous_epoch, *validator_index)?;
+
+            let effective_balance = state.get_effective_balance(*validator_index)?;
+
+            let effective_balance_eth =
+                effective_balance.safe_div(spec.effective_balance_increment)?;
+
+            let mut head_reward = 0u64;
+            let mut target_reward = 0i64;
+            let mut source_reward = 0i64;
+
+            for flag_index in PARTICIPATION_FLAG_WEIGHTS {
+                if eligible {
+                    let voted_correctly = participation_cache
+                        .get_unslashed_participating_indices(flag_index as usize, previous_epoch)
+                        .is_ok();
+                    if voted_correctly {
+                        let total_reward = ideal_rewards_hashmap
+                            .get(&(flag_index, effective_balance_eth, penalty))
+                            .ok_or(BeaconChainError::AttestationRewardsSyncError)?;
+
+                        if flag_index == 0 {
+                            head_reward += total_reward;
+                        } else if flag_index == 1 {
+                            target_reward += *total_reward as i64;
+                        } else if flag_index == 2 {
+                            source_reward += *total_reward as i64;
+                        }
+                    } else if flag_index == 0 {
+                        head_reward = 0;
+                    } else if flag_index == 1 {
+                        target_reward = penalty;
+                    } else if flag_index == 2 {
+                        source_reward = penalty;
+                    }
```

Comment:
These flag indices aren't right (0 = source, 2 = head), lets use the constants from here again:

https://github.com/sigp/lighthouse/blob/38514c07f222ff7783834c48cf5c0a6ee7f346d0/consensus/types/src/consts.rs#L2-L4